src/compiler/scala/tools/nsc/typechecker/Infer.scala

```diff
@@ -823,13 +823,46 @@ trait Infer extends Checkable {
         case _                                                              => onRight
       }
     }
+
     private def isAsSpecificValueType(tpe1: Type, tpe2: Type, undef1: List[Symbol], undef2: List[Symbol]): Boolean = tpe1 match {
       case PolyType(tparams1, rtpe1) =>
         isAsSpecificValueType(rtpe1, tpe2, undef1 ::: tparams1, undef2)
       case _                         =>
         tpe2 match {
           case PolyType(tparams2, rtpe2) => isAsSpecificValueType(tpe1, rtpe2, undef1, undef2 ::: tparams2)
-          case _                         => existentialAbstraction(undef1, tpe1) <:< existentialAbstraction(undef2, tpe2)
+          case _                         =>
+
+            if (settings.YdottySpecificity) {
+              // Backport of fix for https://github.com/scala/bug/issues/2509
+              // from Dotty https://github.com/lampepfl/dotty/commit/89540268e6c49fb92b9ca61249e46bb59981bf5a
+              val flip = new TypeMap(trackVariance = true) {
+                var subst: Map[Symbol, Symbol] = Map.empty
+                def apply(tp: Type): Type = tp match {
+                  case TypeRef(pre, sym, args) if variance > 0 && sym.typeParams.exists(_.isContravariant) =>
+                    val flippedSym =
+                      subst.get(sym) match {
+                        case None =>
+                          val flippedClassSym = sym.owner.newClassSymbol(sym.name.toTypeName, sym.pos, sym.flags)
```

Comment:
Intuitively I would expect `Quux[B]` to be more specific than `Z[A]` for the reasons you gave. The fact that it doesn't in dotty seems like an oversight to me /cc @odersky .